In [48]:
import pandas as pd
import numpy as np

# ML Zoomcamp HW 1

# Question 1
Pandas version

In [2]:
pd.__version__

'2.1.0'

Pandas version 2.1.0

# Gettting the data

In [5]:
!wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv

--2023-09-15 08:37:37--  https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.110.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1423529 (1.4M) [text/plain]
Saving to: ‘housing.csv’

housing.csv         100%[===================>]   1.36M  --.-KB/s    in 0.1s    

2023-09-15 08:37:37 (9.40 MB/s) - ‘housing.csv’ saved [1423529/1423529]



# Question 2

In [34]:
df = pd.read_csv("housing.csv")

In [11]:
df.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           20640 non-null  float64
 1   latitude            20640 non-null  float64
 2   housing_median_age  20640 non-null  float64
 3   total_rooms         20640 non-null  float64
 4   total_bedrooms      20433 non-null  float64
 5   population          20640 non-null  float64
 6   households          20640 non-null  float64
 7   median_income       20640 non-null  float64
 8   median_house_value  20640 non-null  float64
 9   ocean_proximity     20640 non-null  object 
dtypes: float64(9), object(1)
memory usage: 1.6+ MB


In [8]:
df.shape

(20640, 10)

10 columns

# Question 3

In [18]:
df.isnull().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

Total bedrooms has missing values

# Question 4

In [21]:
df.nunique()

longitude               844
latitude                862
housing_median_age       52
total_rooms            5926
total_bedrooms         1923
population             3888
households             1815
median_income         12928
median_house_value     3842
ocean_proximity           5
dtype: int64

Ocean proximity has 5 unique values

# Question 5

In [24]:
df[df["ocean_proximity"]=="NEAR BAY"]["median_house_value"].mean()

259212.31179039303

Average value of houses near the bay 259212

# Question 6

In [35]:
df.isnull().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

In [36]:
total_bedroom_ave = df["total_bedrooms"].mean()
print(total_bedroom_ave)

537.8705525375618


In [37]:
total_bedroom_ave = df["total_bedrooms"].mean()
df.fillna(total_bedroom_ave, inplace=True)
df.isnull().sum()

longitude             0
latitude              0
housing_median_age    0
total_rooms           0
total_bedrooms        0
population            0
households            0
median_income         0
median_house_value    0
ocean_proximity       0
dtype: int64

In [38]:
total_bedroom_ave = df["total_bedrooms"].mean()
print(total_bedroom_ave)

537.8705525375617


# Question 7

Select all the options located on islands.

Select only columns housing_median_age, total_rooms, total_bedrooms.

Get the underlying NumPy array. Let's call it X.

Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.

Compute the inverse of XTX.

Create an array y with values [950, 1300, 800, 1000, 1300].

Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.

What's the value of the last element of w?

In [44]:
island = df[df["ocean_proximity"]=='ISLAND']
island = island[["housing_median_age", "total_rooms", "total_bedrooms"]]

In [50]:
X = island.values
X_transpose = X.T
XTX = np.dot(X_transpose, X)
XTX_inverse = np.linalg.inv(XTX)

In [52]:
y = np.array([950, 1300, 800, 1000, 1300])

In [57]:
A = np.dot(np.linalg.inv(XTX), X.T)
w = np.dot(A, y)

last_w = w[-1]

print(last_w)

5.699229455065565


5.6992